### ISSUE TITLE
Algorithm

### ISSUE THREAD

**Issue by bstemper** [scikit-optimize/scikit-optimize#626]
I'm using the package to do hyper parameter optimisation of a 3-layer neural network. When calling

> res_gp = gp_minimize(train_bayes, space, n_calls=200, n_random_starts=20, acq_func='EIps',
                     n_jobs=-1, verbose=True, random_state=random_seed)

and plotting the res_gp object, the following plot suggests that the optimisation is indeed working
![plot_convergence](https://user-images.githubusercontent.com/16957739/35674450-fed90c22-0744-11e8-8b3a-4d54854c8a6c.jpg)
but then this plot raises some eyebrows:
![plot_objective](https://user-images.githubusercontent.com/16957739/35674548-4801a198-0745-11e8-8a27-8a325cfaf584.jpg)
calls this into question. 

If I am not mistaken (see also issue #624 ), lighter colours correspond to minima, yet the algorithm seems to try to stay away from minimal regions? Is this a bug?

**Comment by iaroslav-ai** [comment]
Hi @bstemper , thanks for reporting this!

You can obtain such unexpected plots with other functions. Consider [example here](https://gist.github.com/iaroslav-ai/42bf1088c129b497dc097ee518d23d84), and plot that it produces:

![figure](https://user-images.githubusercontent.com/14038787/35682262-59637568-0760-11e8-8f50-11c152cf1310.png)

It would seem that there is some issue with convergence and one can do better. Yet if you evaluate manually the best found surrogate, you will see that in fact at best found point surrogate value is minimal. Below is the comparison of estimated objective at best found point, and minimum of objective over 1000000 random samples from search space (you can also find this in code example):

```bash
Best objective value found:
-12.099035547075625
Surrogate value at best x:
-12.098869568765089
Best objective value found by sampling 1000000 random points:
-12.07848016183781
```

Long story short: the plots are misleading. 

There are a few reasons for such behavior. Here are a few:
* Generally the objective is assumed to be sampled from a Gaussian process prior where you can have complex non-linear correlations between n > 2 dimensions, which hence cannot be represented in 2d plots.
* The average over values of objective is plotted (over randomly sampled invisible parameters), whereas a user / algorithm is interested in minimum value of objective.
In addition to all this, for small number of function evaluations surrogate does not accurately represent the “real” objective (but `n_calls` seems sufficient in example and your case to be accurate). 

It would be great if you can use verification code in the example to test this for yourself, if you have the results of optimization saved somewhere. 

You can also test example code with other random seed values to see other examples of such plots.

**Comment by iaroslav-ai** [comment]
(This is the code that I talk about, just replace `search_result` with your results)
```python
search_result = gp_minimize(fnc, dims, n_calls=200)

# get the search space and best x, y, model
space = search_result.space
model = search_result.models[-1]
best_x = search_result.x
best_y = search_result.fun

# sample points from search space
random_samples = np.array(space.rvs(n_samples=1000000))
best_x_sample = np.array([best_x])

# make estimations with surrogate
y_random = model.predict(space.transform(random_samples))
best_y_estimated = model.predict(space.transform(best_x_sample))[0]

# print best found actual objective value
print("Best objective value found:")
print(search_result.fun)

# print the best estimated objective
print("Surrogate value at best x:")
print(best_y_estimated)

# print the minimum objective found by random sampling from parameter space
print("Best objective value found by sampling 1000000 random points:")
print(np.min(y_random))

# visualize plots for comparison
import matplotlib.pyplot as plt
plot_objective(result=search_result)
plt.show()
```

**Comment by bstemper** [comment]
Thanks @iaroslav-ai , I took your snippet exactly as you posted and ran this on my scipy object:

```
Best objective value found:
0.280820012093
Surrogate value at best x:
[ 0.64676332  7.40975677]
Best objective value found by sampling 1000000 random points:
0.430572724824
```

I am running this on scikit-optimize v0.4 and Python 3 by the way if that matters.

**Comment by betatim** [comment]
Should we turn https://gist.github.com/iaroslav-ai/42bf1088c129b497dc097ee518d23d84 into an example notebook for future reference?

**Comment by iaroslav-ai** [comment]
Yup, should be easy to do with the comments and code that I already have.

**Comment by bstemper** [comment]
Are you guys not a bit concerned about the results that I posted above?

**Comment by betatim** [comment]
I will experiment a bit on the partial dependence plotting with some simple 3D functions. I did some testing when we first added it but because it is so tricky to imagine what a projection of a high dimensional function into 1D or 2D should look like, I would not be surprised if there is a bug :-(

**Comment by betatim** [comment]
https://gist.github.com/betatim/febd1b6cb4c21ce1d99cc10dcaa203eb some experiments with simple functions. So far I think they look as you would expect. Though for the last one already it is hard to decide.

**Comment by betatim** [comment]
Added some more examples at the bottom: https://gist.github.com/betatim/febd1b6cb4c21ce1d99cc10dcaa203eb/a0899441dd64bb32778ca803e176bb4ce05ca500 I think especially `tricky_func` shows how the minimum in the projected (1D) view is somewhere else to where the "global" minimum is. The RF based surrogate also shows that your surrogate does not have to build a good model of the function you are minimising. People often get confused when the surrogate looks nothing like the "true" function, but it makes sense if you think that all we are trying to do is find the minimum, not model the function.

**Comment by iaroslav-ai** [comment]
Nice examples, @betatim !

Here is an idea: replace `np.mean` with `np.min` inside `partial_dependence`, [here](https://github.com/scikit-optimize/scikit-optimize/blob/2121e8486722c1021b8e4f0d864a733cbadce0b1/skopt/plots.py#L234) and [here](https://github.com/scikit-optimize/scikit-optimize/blob/2121e8486722c1021b8e4f0d864a733cbadce0b1/skopt/plots.py#L254). Then the global minima and minima on the plots should coincide. The plots then turn into something like "partial minimum", where for different combinations of 1 or 2 arguments you can see what is the minimum expected objective. These should behave a bit more like what you would expect them to do.

**Comment by iaroslav-ai** [comment]
With regards to checking the minimum: 'EIps' is used (time is taken into account in objective minimized), it is a bit non - trivial to check the minimum of a surrogate due to acquisition function used. The code that I posted here does not apply to this scenario. 

What I see thus far also is that `partial_dependence` does not take care of `..ps` like acquisition functions, which might add to the uncertainty :/

**Comment by betatim** [comment]
👍 to the idea of using `np.min`. In these cases it does what you predicted.

Given that a lot of people have not used PDP before and they are kind of hard to read/need some practice ... should we make them subtly different from what PDP normally means by switching to `min` instead of `mean`? It seems like a good idea, but ...

**Comment by iaroslav-ai** [comment]
I am all for switching to `np.min` tbh. Our main interest is in minimizing values anyway :)

I am also fine with renaming these functions to avoid confusion.

**Comment by iaroslav-ai** [comment]
I imagine that such minimum plots can be for instance used when a user wants to select after the optimization other parameter configurations, which somehow suit user needs better. But I would need to think a bit for exact scenario.

**Comment by jonathanng** [comment]
Or rather instead of plotting just the min or mean, plot every decile.

**Comment by ljmartin** [comment]
Just wanted to add a +1 for this - just tried it and the results are far more intuitive even if technically different from the original intended effect. Perhaps we can add the 'minimum' as a kwarg? I'm happy to do this if you like.
